src/xrpld/app/tx/detail/Batch.cpp

```diff
@@ -0,0 +1,250 @@
+            return temARRAY_TOO_LARGE;
+        }
+
+        // Add the batch signers to the set.
+        for (auto const& signer : signers)
+        {
+            AccountID const innerAccount = signer.getAccountID(sfAccount);
+            if (!batchSignersSet.insert(innerAccount).second)
+            {
+                JLOG(ctx.j.trace())
+                    << "Batch: Duplicate signer found: " << innerAccount;
+                return temINVALID_BATCH;
+            }
+        }
+
+        // Check the batch signers signatures.
+        auto const requireCanonicalSig =
+            ctx.rules.enabled(featureRequireFullyCanonicalSig)
+            ? STTx::RequireFullyCanonicalSig::yes
+            : STTx::RequireFullyCanonicalSig::no;
+        auto const sigResult =
+            ctx.tx.checkBatchSign(requireCanonicalSig, ctx.rules);
+
+        if (!sigResult)
+        {
+            JLOG(ctx.j.trace()) << "Batch: invalid batch txn signature.";
+            return temBAD_SIGNATURE;
+        }
+    }
+
+    std::unordered_set<AccountID> uniqueSigners;
+    std::unordered_set<uint256, beast::uhash<>> uniqueHashes;
+    for (int i = 0; i < txns.size(); ++i)
+    {
+        if (!uniqueHashes.emplace(hashes[i]).second)
+        {
+            JLOG(ctx.j.trace()) << "Batch: duplicate TxID found.";
+            return temMALFORMED;
+        }
+
+        STTx const stx = STTx{STObject(txns[i])};
+        if (stx.getTransactionID() != hashes[i])
+        {
+            JLOG(ctx.j.trace()) << "Batch: txn hash does not match TxIDs hash."
+                                << "index: " << i;
+            return temMALFORMED;
+        }
+
+        if (!stx.isFieldPresent(sfTransactionType))
+        {
+            JLOG(ctx.j.trace())
+                << "Batch: TransactionType missing in inner txn."
+                << "index: " << i;
+            return temINVALID_BATCH;  // LCOV_EXCL_LINE
+        }
+
+        if (stx.getFieldU16(sfTransactionType) == ttBATCH)
+        {
+            JLOG(ctx.j.trace())
+                << "Batch: batch cannot have an inner batch txn."
+                << "index: " << i;
+            return temINVALID_BATCH;
+        }
+
+        auto const innerAccount = stx.getAccountID(sfAccount);
+        if (stx.getFieldU16(sfTransactionType) == ttACCOUNT_DELETE &&
+            innerAccount == outerAccount)
+        {
+            JLOG(ctx.j.trace())
+                << "Batch: inner txn cannot be account delete when inner and "
+                   "outer accounts are the same."
+                << "index: " << i;
+            return temINVALID_BATCH;
+        }
+
+        if (auto const preflightResult =
+                ripple::preflight(ctx.app, ctx.rules, stx, tapFAIL_HARD, ctx.j);
+            preflightResult.ter != tesSUCCESS)
+        {
+            JLOG(ctx.j.trace()) << "Batch: inner txn preflight failed."
+                                << "index: " << i;
+            return preflightResult.ter;
```

Comment:
Please add a test for this error.

Reply:
fixed: https://github.com/XRPLF/rippled/pull/5060/commits/d385958c234aa8e81e6238de81b9ed7b7bf1dbf3